### PR TITLE
feat: add standard PPO training with GAE and value critic (Transformers/FSDP)

### PR DESCRIPTION
**Pull request by xxyyrr598**
# PR type
- [x] New Feature
- [ ] Bug Fix
- [ ] Document Updates
- [ ] More Models or Datasets Support

# PR information

## Summary

This PR adds the first **standard PPO + GAE** training pipeline to Twinkle, on top of the existing GRPO/DPO RL stack. It introduces the missing PPO components — a token-level GAE advantage estimator, a full-parameter value critic, and a clipped value loss — and provides an end-to-end GSM8K cookbook with LoRA policy + full-parameter critic + vLLM sampler.

### New components
- **`cookbook/rl/ppo/`** — end-to-end GSM8K PPO training script & shell launcher:
  - LoRA policy (FSDP) + full-parameter critic (FSDP) + vLLM sampler on separate GPU groups
  - Frozen base model as reference policy via `forward_only(disable_lora=True)` (no extra ref GPUs)
  - `ppo_epochs` mini-batch updates over fixed rollouts
- **`src/twinkle/advantage/gae.py`** — `GAEAdvantage`, token-level GAE with:
  - terminal bootstrap = 0, padding masked out
  - per-token KL shaping reward (`logp_old - logp_ref`) with `kl_coef`
  - optional advantage normalization
- **`src/twinkle/loss/value.py`** — `PPOValueLoss`, clipped value objective `max((v-r)², (clip(v,v_old)-r)²)`
- **`src/twinkle/model/transformers/value_model.py`** — `TransformersValueModel`, causal-LM backbone with a `hidden_size -> 1` value head replacing the LM head
- **`src/twinkle/metric/ppo.py`** — `PPOValueMetric` (value/return mean, value clip ratio, explained variance)
- **`PPOLoss(GRPOLoss)` / `PPOMetric(GRPOMetric)`** — semantic aliases reusing the existing clipped policy objective & ratio/clip statistics

### Framework generalization (backward-compatible)
- `ModelOutput` gains a `values` field; forward/forward_only can emit `values` when the loss declares `require_values`
- `processor` packing, `nccl_safe` wrapper, and `loss/base` handle the new `values` tensor
- New CLI args: `--critic-model-gpus`, `--ppo-epochs`, `--gamma`, `--gae-lambda`, `--kl-coef`, `--normalize-advantages`, `--critic-learning-rate`, `--value-clip`

### Tests
- `tests/advantage/test_gae.py` — single/multi-step GAE, terminal bootstrap, padding mask, normalization, sparse reward + KL shaping
- `tests/loss/test_ppo.py` — `PPOLoss` parity with `GRPOLoss`, clipped/unclipped value loss, ragged token alignment
- `tests/model/test_value_model.py` — value-head forward/backward, save/load
- `tests/cli/test_cli.py` — new PPO CLI args

### Docs
- README / README_ZH cookbook table updated with the PPO entry

### Results
explained variance
KL Clip Ratio
reward

## Usage

```bash
cd cookbook/rl/ppo
bash ppo.sh   # default 12 GPUs: 4 policy + 4 critic + 4 sampler

### CODE REVIEW

File: src/twinkle/model/transformers/transformers.py [hunk 1]
```diff
@@ -504,6 +505,9 @@ def forward(self, *, inputs: Union[InputFeature, List[InputFeature], List[Trajec
             else:
                 outputs['logps'] = selective_log_softmax(logits, masked_labels)
             del logits
+        if loss_require_values:
+            values = outputs['logits']
+            outputs['values'] = values.squeeze(-1) if values.shape[-1] == 1 else values
```
Comment (hjh0119) [inline src/twinkle/model/transformers/transformers.py:510]: 确认下这里是否要做shift移位

File: src/twinkle/infra/_ray/ray_helper.py [hunk 2]
```diff
@@ -71,7 +71,24 @@ def initialize(nproc_per_node: int, ncpu_proc_per_node: int, device_groups: List
...
+            if len(_non_cpu_types) == 1:
+                _dev_type = next(iter(_non_cpu_types))
+                if _dev_type == 'NPU':
+                    try:
+                        import torch
+                        _npu_count = torch.npu.device_count() if hasattr(torch, 'npu') else 0
+                        if _npu_count > 0:
+                            _resources['NPU'] = float(_npu_count)
+                    except Exception:
+                        pass
```
Comment (hjh0119) [inline src/twinkle/infra/_ray/ray_helper.py]: 单独抽象成一个方法比较好

File: src/twinkle/cli/cli.py [hunk 3]
```diff
@@ -205,6 +207,12 @@ class RLArgs:
     gkd_temperature: float = 1.0
     gkd_topk: int = 64
     router_replay_mode: Literal['disabled', 'R2', 'R3'] = 'disabled'
+    ppo_epochs: int = 4
```
Comment (hjh0119) [inline src/twinkle/cli/cli.py]: 是不是可以复用 TrainArgs.num_train_epochs

File: .gitattributes [hunk 4]
```diff
@@ -0,0 +1 @@
+* text=auto
```
Comment (hjh0119) [inline .gitattributes:1]: 这里是为什么

**Review by hjh0119 — CHANGES_REQUESTED** [review]

File: .gitattributes (hunk 4 above)
Comment (xxyyrr598) [reply .gitattributes:1]: 这个改动纯粹是行尾规范化，不影响任何运行逻辑，整个文件只有一行：

作用：

* 匹配所有文件；
text=auto 让 Git 自动区分文本/二进制文件。文本文件入库时统一按 LF 存储（无论提交者是什么平台），检出时再按当前平台转回（Windows 用 CRLF，Linux/macOS 用 LF）；二进制文件完全不碰。
为什么加：

我在 Windows/WSL 下开发，而 CI 跑在 Linux 上。不做规范化的话，只改一行代码就可能因为行尾（CRLF vs LF）变化导致整个文件显示成全量修改，污染 diff、干扰 review。
同时避免把 CRLF 意外提交进仓库，防止 Linux CI 上出现由 \r 引起的偶发失败（比如 shell/python 文件）。
它只影响 Git 对后续提交的行尾处理，不改变任何代码、依赖或行为。如果维护者认为这个 PR 不需要包含该规范化，我可以移除。

File: src/twinkle/model/transformers/transformers.py (hunk 1 above)
Comment (xxyyrr598) [reply src/twinkle/model/transformers/transformers.py:510]: 此处无需对 values 额外 shift。Twinkle 的 labels 已遵循 causal-LM 的预移位约定，logits[t] 直接与 labels[t] 对齐；Value Model 的 values[t] 也表示预测该 token action 前状态的 V(s_t)。PPO 中 logprob、value、reward 均按同一 labels mask 选取，GAE 所需的 V(s_{t+1}) 在 advantage 计算阶段通过 next value 获取，末位置 bootstrap 为 0。若在此处单独 shift values，会破坏该对齐关系。

File: src/twinkle/infra/_ray/ray_helper.py (hunk 2 above)
Comment (xxyyrr598) [reply src/twinkle/infra/_ray/ray_helper.py]: NPU 资源探测逻辑抽取为 RayHelper._get_ray_custom_resources，使 initialize() 只负责 Ray 初始化。  
该方法仅在 device group 为 NPU 时通过 torch.npu.device_count() 注册 {"NPU": count}；GPU 继续由 Ray 原生自动发现，因此不会引入额外的 GPU custom resource。对于 CPU、GPU 或当前不支持的混合设备组，方法返回空字典，保持原有行为。  
这里的资源注册用于当前进程启动本地 Ray 的场景；若连接已启动的 Ray 集群，则仍需在各节点启动 Ray 时注册 NPU resource。

File: src/twinkle/cli/cli.py (hunk 3 above)
Comment (xxyyrr598) [reply src/twinkle/cli/cli.py]: 已经修改成使用TrainArgs.num_train_epochs

File: src/twinkle/loss/grpo.py [hunk 5]
```diff
@@ -304,6 +304,15 @@ def __call__(
         return LossOutput(loss=loss, num_tokens=0)
 
 
+class PPOLoss(GRPOLoss):
```
Comment (meichangsu1) [inline src/twinkle/loss/grpo.py:307]: 不需要重写_aggregate_loss 吗

Comment (xxyyrr598) [reply src/twinkle/loss/grpo.py:307]: 确认后，PPO 的 policy loss 更适合默认按所有有效 response token 直接求平均，即 token-mean。这样每个 action token 权重一致，也与 PPOValueLoss 当前的 token-level 聚合方式保持一致。
同时在 PPOLoss 中保留了 seq-mean-token-mean 模式，即先对每条 response 内的 token loss 求平均，再对 batch 求平均。该模式能保证不同长度的 response 具有相同总权重，但会使短 response 中单个 token 的权重更大。
目前这个选择仅用于 PPO，因此没有添加到公共 CLI 参数中，而是在 PPO cookbook 中显式指定 loss_agg_mode='token-mean'。等后续其他算法也需要统一选择聚合方式时，再考虑将其抽象为公共配置。两种聚合模式均已补充单元测试。

**Review by tpx818 — APPROVED** [review]

**Review by hjh0119 — APPROVED** [review]
